enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+Node Health Check Enhancement 
+
+---
+title: node-healthcheck-controller
+authors:
+  - "@rgolangh"
+reviewers:
+  - "@abeekof"
+  - "@n1r1"
+  - "@mshitrit"
+  - "@slintes"
+  - "@dhellmann"
+approvers:
+  - "@abeekof"
+  - "@n1r1"
+---
+
+# Node health checking
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+
+Enable opt in for automated node health checking and create objects
+for unhealthy nodes, to be handled later by external controllers
+that comply with the existing  [api].
+
+## Motivation
+
+Enable clusters that do not have a functioning Machine API to detect and respond to node-level failures in order to recover workloads , and/or restore cluster capacity.
+
+The existing Machine Health Check controller requires a functioning machine-api. This prevents its use in the majority of Kubernetes and OpenShift clusters.
+The proposed Node Healthcheck Controller (NHC) provides a generic solution that fits all installation types, together with an external remediation API that allows a pluggable remediation system to be installed and remediate unhealthy nodes.
+
+### Goals
+
+- Create an installation-type agnostic node health check solution
+
+- Feature parity with Machine Health Check controller
+
+- Providing a default strategy combined with [poison-pill].
+
+- Endorse usage of external [remediation api]
+
+- Allow defining different unhealthy criteria by node selectors, and have a sane strategy for handling nodes that match multiple NHC CRs.
+
+- Provide defaults health checks and default remedy custom resource creation.
+
+- Long term - Make Machine Health Check controller only implement the external remediation api and drop the default Machine API based remediation mechanism
+
+
+### Non-Goals
+
+- Coordination of remedy operation - operators/controllers implementers
+expected to track the [node maintenance lease proposal]
+
+- Recording long-term stable history of all health-check failures or remediation
+  actions.
+
+- Detection of Machines that never become healthy Nodes is not part of the initial implementation.  Machine API based cluster can/should continue to use MHC at this time.
+
+## Proposal
```

Comment:
It would be useful to have a subsection describing how the feature provided by the new controller is used, end to end. Something like

1. User installs the NodeHealthCheck operator
2. User creates a NodeHealthCheck resource describing rules for monitoring nodes. (How do they decide what rules apply? Is there more configuration to do?)
3. The controller sees that a node fails the check rule described.
4. The controller does ... (I think it's creating instructions for a remediation controller?)
5. The remediation controller does ...
6. The node is deleted/rebooted/reprovisioned/other

Reply:
Ack